### PR TITLE
Operators Agents

### PR DESCRIPTION
**Pull request by gambol99**
- adding operator subscriptions controller and views
- adding operator agent
- adding the manage subscriptions button to the kubernete integretion
- bumping the version of the hub-cluster-creator to v0.1.0
- installed the cluster dependencies via the OLM
- adding the first draft for the ui to manage olm subscriptions and updates

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -101,12 +101,12 @@ GEM
...
       erubi (>= 1.0.0)
       rack (>= 0.9.0)
-    bindex (0.7.0)
+    bindex (0.8.1)
     binding_of_caller (0.8.0)
       debug_inspector (>= 0.0.1)
     bootsnap (1.4.5)
       msgpack (~> 1.0)
-    bootstrap_form (4.2.0)
+    bootstrap_form (4.3.0)
```
Comment (jits) [inline Gemfile.lock]: Some of these gem upgrades look like they should be part of a separate PR (and potentially for another time)?

Comment (gambol99) [reply Gemfile.lock]: Admittedly I only wanted to update the hub-clusters-creator tag and reference when i did a `bundle update` ... I'm happy revert and keep just the https://github.com/appvia/appvia-hub/pull/300/files#diff-e79a60dc6b85309ae70a6ea8261eaf95R3?

Comment (jits) [reply Gemfile.lock]: Yeah, just need to bump the version in `Gemfile` then run `bundle` (`bundle install`)

File: app/controllers/admin/integrations_controller.rb [hunk 2]
```diff
@@ -11,6 +13,7 @@ def index
       @group_to_expand = params[:expand]
       @unmask = params.key? 'unmask'
 
+      @agents = AgentsService
```
Comment (jits) [inline app/controllers/admin/integrations_controller.rb]: Doesn't look this ends up getting used anywhere?

File: app/controllers/admin/integrations_controller.rb [hunk 3]
```diff
@@ -48,6 +51,59 @@ def edit
       @potential_teams = find_potential_teams @integration.provider_id
     end
 
+    # GET /admin/integrations/:integration_id/operators/:id/subscriptions/:namespace/:name/approve
```
Comment (jits) [inline app/controllers/admin/integrations_controller.rb]: I think all this new functionality should live in a separate controller, as this controller is already massive enough. You can still keep the same routes, just that they should be set up to use actions from a new controller for all this new functionality.

File: app/controllers/admin/integrations_controller.rb [hunk 4]
```diff
@@ -48,6 +51,59 @@ def edit
...
+    def approve_subscription
+      name = params[:name]
+      namespace = params[:namespace]
+
+      # approve the subscription
+      @agent.approve_subscription(namespace, name)
+
+      # invalidate the cache
+      # Rails.cache.delete(params[:integration_id])
```
Comment (jits) [inline app/controllers/admin/integrations_controller.rb]: Could you remove unused / commented out code like this?

File: app/controllers/admin/integrations_controller.rb (hunk 3 above)
Comment (jits) [reply app/controllers/admin/integrations_controller.rb]: Additionally, could you wrap most of this operators/subscriptions logic in a new _service_, and then call the service from the controller/views/etc. – see for example https://github.com/appvia/appvia-hub/blob/master/app/services/service_catalog_service.rb

File: app/views/admin/integrations/_card.html.erb [hunk 5]
```diff
@@ -70,8 +70,12 @@
           unmask: unmask
         }
     %>
+
```
Comment (jits) [inline app/views/admin/integrations/_card.html.erb]:
```suggestion
```

File: app/workers/admin/task_worker.rb [hunk 6]
```diff
@@ -70,7 +70,7 @@ class TaskWorker
           'loki' => loki_integration.id
         }
 
-        if results[:services][:aws_sb]
+        if results[:services].key?(:catalog) && results[:services][:catalog][:enabled]
```
Comment (jits) [inline app/workers/admin/task_worker.rb]: This could be something to pull out in a separate PR and merge in asap, as it's bug in `master`?

Comment (gambol99) [reply app/workers/admin/task_worker.rb]: https://github.com/appvia/appvia-hub/pull/301

File: app/controllers/admin/integrations_controller.rb (hunk 3 above)
Comment (gambol99) [reply app/controllers/admin/integrations_controller.rb]: So moved the subscriptions into it's own controller .. Working out how to right specs now ..

**Comment by jits** [comment]
@gambol99 –

> @jits ... can have a quick review .. I'll try and move into an a service .. but will probably check into another PR as this one is getting hefty .. (i.e agent_only_olm...agent_service)

Sure – will wait until that's done to continue reviewing this (as I expect this will change as a result). Note also that by wrapping the logic in a separate service, your request specs for the new endpoints can be much easier, as you can mock out the calls to the service in those request specs. Ideally, the service itself would then have it's own specs (where you mock out the boundaries below, where needed).

**Comment by gambol99** [comment]
Merged in the subscription service and added an additional spec for it

File: app/agents/operator_agent.rb [hunk 7]
```diff
@@ -0,0 +1,227 @@
+# rubocop:disable Metrics/ClassLength
+class OperatorAgent
+  # is the main api version
+  API_V1 = 'operators.coreos.com/v1'.freeze
+  # the alpha operator api
+  API_ALPHA_V1 = 'operators.coreos.com/v1alpha1'.freeze
+  # the package api
```
Comment (jits) [inline app/agents/operator_agent.rb]: Are these comments necessary? They are literally describing what the variable name says :D (urgh Golang habits :P)

File: app/agents/operator_agent.rb [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+  PACKAGE_API_V1 = 'packages.operators.coreos.com/v1'.freeze
+
+  def initialize(url:, token: nil, ca_cert: nil, client_certificate: nil, client_key: nil)
+    verify_tls = false
+    verify_tls = true unless ca_cert.nil?
```
Comment (jits) [inline app/agents/operator_agent.rb]: [nitpick] Can just collapse these two lines into `verify_tls = ca_cert.present?`

File: app/controllers/admin/subscriptions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
+module Admin
+  class SubscriptionsController < BaseController
```
Comment (jits) [inline app/controllers/admin/subscriptions_controller.rb]: For this new controller and the new service, seems like calling just `SubscriptionsX` is a bit too generic? Maybe `OperatorSubscriptionsX` is clearer? (as "subscriptions" has so many overloaded meanings in tech).

File: app/controllers/admin/subscriptions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+      @service.approve(namespace, name)
+
+      redirect_to "#{admin_subscriptions_path}/#{@integration.id}", notice: 'Upgrade has been approved and will be upgraded in background'
+    end
+
+    private
+
+    def find_integration
+      @integration = Integration.find params[:subscription_id]
+      @agent = AgentsService.get('operator', @integration.config)
```
Comment (jits) [inline app/controllers/admin/subscriptions_controller.rb]: What happens if the integration is not a `kubernetes` one? Maybe we should short circuit with an error response before carrying on with this line?

File: app/agents/operator_agent.rb [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+    status = subscription.status
+    raise ArgumentError, "subscription, name: #{name}, namespace: #{namespace} has no status" if status.nil?
+    raise ArgumentError, "subscription, name: #{name}, namespace: #{namespace} has no state" if status.state.nil?
+
+    state = status.state
+    return unless state.casecmp('upgradepending').zero?
+
+    installplan_name = status.installplan.name
+    installplan = get_installplan(namespace, installplan_name)
+    raise ArgumentError, "installplan, name: #{installplan_name}, namespace: #{namespace} does not exist" if installplan.nil?
```
Comment (jits) [inline app/agents/operator_agent.rb]: Doesn't seem like `ArgumentError` is the right error to return here and above, as it's nothing to do with a bad argument being passed in?

File: app/controllers/admin/subscriptions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+      model = @service.get(namespace, name)
+
+      render locals: {
+        catalog: model[:catalog],
+        channel: model[:channel],
+        info: model[:info],
+        package: model[:package],
+        subscription: model[:subscription]
+      }
```
Comment (jits) [inline app/controllers/admin/subscriptions_controller.rb]: The "Rails" way to set data for the views is to set `@` instance variables, which are then available in the view. I think setting locals like this could be problematic and break some assumptions down the line.

File: app/services/subscriptions_service.rb [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+  # extract is a helper method to aid in extracting values from the resource
+  # rubocop:disable Lint/RescueException,Lint/HandleExceptions
+  def extract(default_value = 'unknown', &block)
+    begin
+      value = yield block
+      unless value.nil?
+        return value unless value.class == String.class
+        return value unless value.empty?
+      end
+    rescue Exception => _e; end
```
Comment (jits) [inline app/services/subscriptions_service.rb]: Rescue'ing all `Exception` instances is not advised… see:
- https://blog.appsignal.com/2016/10/18/ruby-magic-exceptions-primer.html
- https://ruby-doc.org/core-2.5.7/Exception.html

Comment (jits) [reply app/services/subscriptions_service.rb]: Ah I see you're ignoring the Rubocop `Lint/RescueException` and `Lint/HandleExceptions` cops – why?

File: app/views/admin/subscriptions/_card.html.erb [hunk 14]
```diff
@@ -0,0 +1,57 @@
+<%-
+  base_url = "#{admin_subscriptions_path}/#{@integration.id}"
```
Comment (jits) [inline app/views/admin/subscriptions/_card.html.erb]: Same here (and elsewhere) – don't need to manually construct URLs like this.

**Comment by jits** [comment]
One concern is the amount being rendered in the new page(s), e.g. all the hidden dialog HTML, etc. May be fine in real life usage, but something to keep an eye on in the future.

File: app/views/admin/subscriptions/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+</br></br>
+<%- end -%>
+
+<div class="card">
+  <div class="card-body">
+    <h5 class="card-title">
+      <i class="fas fa-shield-alt"></i><strong>Installed</strong>
+    </h5>
+    <h6 class="card-subtitle mb-2 text-muted">Provides a list of subscriptions which are currently installed in your cluster.</h6>
+    <%- cache "hello" do -%>
```
Comment (jits) [inline app/views/admin/subscriptions/index.html.erb]: Eh?

File: config/routes.rb [hunk 16]
```diff
@@ -5,6 +5,11 @@
   namespace :admin do
     resources :integrations, except: %i[show destroy]
     resource :settings, only: %i[show update]
+    resources :subscriptions do
+      get '/', to: 'subscriptions#index'
+      get '/:namespace/:name', to: 'subscriptions#show'
+      get '/:namespace/:name/approve', to: 'subscriptions#approve'
+    end
```
Comment (jits) [inline config/routes.rb]: (following on from Slack discussion) Try these routes:

```ruby
resources :operator_subscriptions, only: [:index] do
      member do
        get '/:namespace/:name', action: :show, as: ''
        get '/:namespace/:name/approve', action: :approve, as: 'approve'
      end
    end
```

… and now you can use the following URL helpers (note: you don't need the `Rails.application.routes.url_helpers` bit when using them in controllers, views and helpers):

```ruby
[2] [appvia_hub][development] pry(main)> Rails.application.routes.url_helpers.admin_operator_subscriptions_path
=> "/admin/operator_subscriptions"

[3] [appvia_hub][development] pry(main)> Rails.application.routes.url_helpers.admin_operator_subscription_path(1, namespace: 'foo', name: 'bar')
=> "/admin/operator_subscriptions/1/foo/bar"

[4] [appvia_hub][development] pry(main)> Rails.application.routes.url_helpers.approve_admin_operator_subscription_path(1, namespace: 'foo', name: 'bar')
=> "/admin/operator_subscriptions/1/foo/bar/approve"
```

File: app/services/subscriptions_service.rb (hunk 13 above)
Comment (gambol99) [reply app/services/subscriptions_service.rb]: This was a helper method; extracting the values that may or may not be there became a pain as it produce of if statement .. the extract() cleaned it up .. I'll reduce the scope to `NoMethodError, TypeError`

File: config/routes.rb (hunk 16 above)
Comment (gambol99) [reply config/routes.rb]: ARRRRGRGGRRHHH .... sorry that as the ruby scream!! ... so i used the above routing but was hitting errors in difference scenarios so had to keep the https://github.com/appvia/appvia-hub/pull/300/files#diff-4a0e0b66561d51242534db2d2e1acc9fR21 ..

Comment (jits) [reply config/routes.rb]: What are these "different" scenarios? The routes I posted above, together with how the URL helpers should be used, should suffice for all your cases?

Comment (gambol99) [reply config/routes.rb]: Used the above, but they lost the ability to call via `admin/operator_subscriptions/:id` .. added a entry, but complained about a duplicate, tried to rearrange but then fell into the a url with a dot

Comment (jits) [reply config/routes.rb]: Ah OK, that's because of the overloaded nature of the `show` action (i.e. confusing `show` vs `index`) – I see this is not intended to be a resource-ful set of routes; I think I get now what you're after – here's an alternative for you, more in-line with what you already have:

**Routes**

```ruby
scope '/operator_subscriptions' , controller: 'operator_subscriptions' do
  get '/:integration_id/', action: :index, as: 'operator_subscriptions'
  get '/:integration_id/:namespace/:name', action: :show, as: 'operator_subscription'
  get '/:integration_id/:namespace/:name/approve', action: :approve, as: 'approve_operator_subscription'
end
```

… which generates the following route definitions:

```
       admin_operator_subscriptions GET    /admin/operator_subscriptions/:integration_id(.:format)                             admin/operator_subscriptions#index
        admin_operator_subscription GET    /admin/operator_subscriptions/:integration_id/:namespace/:name(.:format)            admin/operator_subscriptions#show
admin_approve_operator_subscription GET    /admin/operator_subscriptions/:integration_id/:namespace/:name/approve(.:format)    admin/operator_subscriptions#approve
```

… with URL helpers:

```ruby
> include Rails.application.routes.url_helpers

> admin_operator_subscriptions_path(123)
=> "/admin/operator_subscriptions/123"

> admin_operator_subscription_path(123, 'foo', 'bar')
=> "/admin/operator_subscriptions/123/foo/bar"

> admin_approve_operator_subscription_path(123, 'foo', 'bar')
=> "/admin/operator_subscriptions/123/foo/bar/approve"
```

… the key things being:

1. should not be using `resources :foo` because that exposes 7 RESTful endpoints that you don't need
1. use named URL helpers to generate URLs (instead of constructing manually)
1. [bonus] `integration_id` is now a more explicit parameter

Hope that works now!

Comment (gambol99) [reply config/routes.rb]: Awesome sauce :-) ... https://github.com/appvia/appvia-hub/pull/300/commits/cb247fdc2d624a6ca45b02d13d693c2974952810

File: app/controllers/admin/operator_subscriptions_controller.rb [hunk 17]
```diff
@@ -0,0 +1,46 @@
+module Admin
+  class OperatorSubscriptionsController < BaseController
+    before_action :find_integration
+
+    authorize_resource class: false
+
+    # GET /admin/subscriptions/:id
```
Comment (jits) [inline app/controllers/admin/operator_subscriptions_controller.rb]: This needs updating to reflect the new routes (and below).

File: app/controllers/admin/operator_subscriptions_controller.rb [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+      @service.approve(namespace, name)
+
+      redirect_to admin_operator_subscriptions_path @integration, notice: 'Upgrade has been approved and will be upgraded in background'
+    end
+
+    private
+
+    def find_integration
+      @integration = Integration.find params[:integration_id]
+      raise ArgumentError, "integrations: #{@integration.id} is not kubernetes" unless @integration.kubernetes?
```
Comment (jits) [inline app/controllers/admin/operator_subscriptions_controller.rb]:
```suggestion
    unless @integration.kubernetes?
      Rails.logger.warn "Integration '#{@integration.name}' (ID: #{@integration.id}) needs to be for provider 'kubernetes' in order for this route to work"
      unprocessable_entity_error
      return
    end
```

File: app/controllers/admin/operator_subscriptions_controller.rb [hunk 19]
```diff
@@ -0,0 +1,46 @@
+module Admin
+  class OperatorSubscriptionsController < BaseController
+    before_action :find_integration
+
+    authorize_resource class: false
+
+    # GET /admin/operator_subscriptions/:id
```
Comment (jits) [inline app/controllers/admin/operator_subscriptions_controller.rb]:
```suggestion
    # GET /admin/operator_subscriptions/:integration_id
```

File: app/controllers/admin/operator_subscriptions_controller.rb [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+    def index
+      @subscriptions = @service.list
+    end
+
+    # GET /admin/operator_subscriptions/:id/:namespace/:name
```
Comment (jits) [inline app/controllers/admin/operator_subscriptions_controller.rb]:
```suggestion
    # GET /admin/operator_subscriptions/:integration_id/:namespace/:name
```

File: app/controllers/admin/operator_subscriptions_controller.rb [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+      model = @service.get(namespace, name)
+
+      @catalog = model[:catalog]
+      @channel = model[:channel]
+      @info = model[:info]
+      @package = model[:package]
+      @subscription = model[:subscription]
+    end
+
+    # GET /admin/operator_subscriptions/:id/:namespace/:name/approve
```
Comment (jits) [inline app/controllers/admin/operator_subscriptions_controller.rb]:
```suggestion
    # GET /admin/operator_subscriptions/:integration_id/:namespace/:name/approve
```

File: spec/requests/admin/operator_subscriptions_spec.rb [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+        }
+      )
+    end
+
+    it_behaves_like 'unauthenticated not allowed' do
+      before do
+        get admin_operator_subscription_path @integration
+      end
+    end
+  end
```
Comment (jits) [inline spec/requests/admin/operator_subscriptions_spec.rb]: Need the additional cases for non-hub admins and hub admins access. (Am assuming that non-hub admins should not be able to access these new endpoint, at the mo).

**Comment by jits** [comment]
Also, may be worth squashing all these new commits into one, before merging in to `master`?

File: spec/requests/admin/operator_subscriptions_spec.rb (hunk 22 above)
Comment (gambol99) [reply spec/requests/admin/operator_subscriptions_spec.rb]: Yep ... I'll probably have to check with you on this one .. Because hitting the index action effectively speaks to a k8s cluster i didn't know how to reference the controller to mock the implementation

Comment (jits) [reply spec/requests/admin/operator_subscriptions_spec.rb]: Now that you have that logic wrapped in a service, you can just mock the calls made to the service from the controller code. See for example: https://github.com/appvia/appvia-hub/blob/f9d9ec81b60dfd7bd63e0f3156246246e4898b42/spec/requests/resources_spec.rb#L411-L490

Comment (jits) [reply spec/requests/admin/operator_subscriptions_spec.rb]: Ahh, I guess this didn't make it in?
